cluster-up/cluster/kind-k8s-sriov-1.17.0/config_sriov.sh

```diff
@@ -334,4 +334,10 @@ wait_allocatable_resource $SRIOV_NODE "openshift.io/$resource_name" $NODE_PF_NUM
 _kubectl get nodes
 _kubectl get pods -n $SRIOV_OPERATOR_NAMESPACE
 echo
+
+# Upon cluster start we might see "etcdserver: request timed out"
+# when accessing API, due to IO pressure.
+# Let the transient pass away (skip it by export SKIP_SRIOV_SLEEP=1)
+[ -z $SKIP_SRIOV_SLEEP ] && echo "sleeping 120 seconds to stabilize etcd, export SKIP_SRIOV_SLEEP=1 to skip"  && sleep 120
```

Comment:
Isn't there a better condition we can check besides just an arbitrary sleep?

Reply:
I am not sure yet beside the info in the PR description, unless we will "wait" for more specific conditions on each pod,
some of the pods or SRIOV operator are known to missing few conditions already, this is why the SRIOV waits for taint addon and then absent.

Please see the PR description 2 references that shows its a known issue that can be caused by IO stress, or if there aren't enough inotify in the system.

Reply:
So what exactly is the issue here? That the etcd cannot get reached by the sriov operator? Or that the etcdserver takes a long time to get up? Or that accessing the cluster API times out, even though the etcd is okay?

If the etcd itself is the problem, can't we rather try to check that? I guess that the operator is missing some kind of state then?

Reply:
Issue is that after cluster is freshly created, first test might experience etcd server timeout when trying to reach the API ep.
etcd is ok, its legit that it happens (see please the Ref links, as its a known issue)

https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4511/pull-kubevirt-e2e-kind-1.17-sriov/1325727568359329793

I saw it several times on different providers in the past, always it was a transient state on a freshly created cluster,
because all the pods starting, we check the conditions that are enough for us (unless we find another specific condition we should check), but then there are couple pods that still access relatively extensively the etcd, causing this pressure.

Reply:
One more thing to make sure is that we dont have a leader change in SRIOV, 
leader change can cause it, and sriov operator has leader election mechanism

Reply:
```
2020-11-09 10:02:48.075278 W | etcdserver: timed out waiting for read index response (local node might have slow network)
2020-11-09 10:02:48.075402 W | etcdserver: read-only range request "key:\"/registry/namespaces/default\" " with result "error:etcdserver: request timed out" took too long (7.184676107s) to execute
2020-11-09 10:02:48.075444 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kubevirt/virt-operator\" " with result "error:etcdserver: request timed out" took too long (7.489442621s) to execute
2020-11-09 10:02:48.075493 W | etcdserver: read-only range request "key:\"/registry/kubevirt.io/virtualmachineinstances/kubevirt-test-default1/testvmicvlq7tmjcblw5bhhtqpzvcjrc87z494xqxnlks9jxc7b4mpb\" " with result "error:etcdserver: request timed out" took too long (8.490783071s) to execute
```
from the time of the error in the test

Reply:
@oshoval what would be creating pressure on the API at this stage? It looks that exactly at the point where you add the  wait, the cluster is already stabilized (all pods ready, all conditions met, and so forth). This means that the pressure should be lower now than during all the deploy steps. This in turn means that all our `kubectl` calls from above worked without a single failure on etcd. This wait looks a little bit questionable.

Reply:
I saw in several providers, that even when the provider finishes, we might still have some transient symtoms
and only on a fresh created provider.
For example https://github.com/kubevirt/kubevirt/issues/4158
Which fails only when runs on check-for-flakes lane, which run the test as the first test and not one in the middle of all the e2e ones.

I believe that even cluster-up is finished, there are some processes like leader election and so on, when all the new pods are started, they are operational but they little more busy at the begining
for example the logs that are mentioned here
https://github.com/etcd-io/etcd/issues/11809#issuecomment-621004811

see please that the link mentioned on the PR description always fails on the first test when it fails,
its a freshly cluster.

What about trying it empirically by merging the PR, see if it removes the flakes and then rethink ?
I already started digging a bit, but don't have the exact solution yet.

Reply:
> I saw in several providers, that even when the provider finishes, we might still have some transient symtoms
> and only on a fresh created provider.

On which providers you saw that happen? could you refer to build log?

> For example [kubevirt/kubevirt#4158](https://github.com/kubevirt/kubevirt/issues/4158)
> Which fails only when runs on check-for-flakes lane, which run the test as the first test and not one in the middle of all the e2e ones.

The job that this issue refers to does not have ectd time out errors, I am not sure it is good example

> I believe that even cluster-up is finished, there are some processes like leader election and so on.
> When all the new pods are started, they are operational but they little more busy at the beginning,
> for example the logs that are mentioned here
> [etcd-io/etcd#11809 (comment)](https://github.com/etcd-io/etcd/issues/11809#issuecomment-621004811)
> 

Can you back this with reproducer or logs you saw?

> see please that the link mentioned on the PR description always fails on the first test when it fails,
> its a freshly cluster.
> 

Note that it could happens on earlier stage, on [CSR for creating a certificate](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4444/pull-kubevirt-e2e-kind-1.17-sriov/1326121036692852736) or when creating a secret which happens right after the certificate is ready.

> What about trying it empirically by merging the PR, see if it removes the flakes and then rethink ?
> I already started digging a bit, but don't have the exact solution yet.

This should be verified to actually affect before merge

> EDIT:
> about "This in turn means that all our kubectl calls from above worked without a single failure on etcd. This wait looks a little bit questionable."
> Just a note - See please that in some cases we had to add retry (see for example the CSR WIP PR #472 that suffer from etcd timeout), but we indeed care from the point after the cluster-up is "done", and its affect on e2e tests.

About Sriov operator leader election, I verified and we use only one replica so the leader election shouldn't affect

Reply:
1. See please the issue i enclosed in the above message, 
https://github.com/kubevirt/kubevirt/issues/4158,
even that pods do exists, there is a hiccup and the node list is empty, i dont have the specific reason, but its a hiccups, i saw those on more providers in the past, for example ocp with 3 masters and there it was exactly the "and" of those:
A. on startup
B. due to disk pressure
C. etcd timeout
And saw more, i think i created once a PR that fixed cluster-sync due to hiccups on other lane beside the 3 masters PR.

2. please see 1, its an hiccup, and etcd is the natural reason for it usually from all the stuff i saw in the past.
3. not have at hand.
4. in the tests its after CSR is done, if CSR had error it wont finish cluster-up
5. see please https://github.com/kubevirt/kubevirtci/pull/473#issuecomment-724808265 the first answer here.

Reply:
see this one that had the same problem (see PR description) https://github.com/kubevirt/kubevirt/pull/3095
at the end of kubevirt `make cluster-sync`

Reply:
see this one that had the same problem (see PR description) kubevirt/kubevirt#3095
at the end of kubevirt make cluster-sync
(referred to it in [1] above).
We dont need to care if its the same string, hiccup is hiccup thats what matters, most of them have things in common, we should understand if its the family of them imo.

Moreover << request time out itself is mentioned in the comments there.

Reply:
about 2 
opened this draft to try and catch the error that happens
https://github.com/kubevirt/kubevirt/pull/4523

Reply:
btw its interesting (above message), it might be after cluster-sync what we see on e2e.....

about 2
opened this draft to try and catch the error that happens
kubevirt/kubevirt#4523